// @algolia/predict-core/dist/esm/tools/props.js
// @TODO: Installing `@algolia/predict-tools` as a dev dependency to retrieve
// the `PredictToolsProps` type creates Rollup cycle issues:
// "@algolia/predict-core -> @algolia/predict-tools -> @algolia/predict-core"
// We should be able to find a way to install it without cycle issues.
export function getPredictToolsEventProps(sendMessage) {
  return {
    onHighlightZone: function onHighlightZone(zoneId, params) {
      sendMessage({
        type: 'request-highlight-zone',
        data: {
          zoneId: zoneId,
          params: params
        }
      });
    },
    onUnhighlightZone: function onUnhighlightZone(zoneId) {
      sendMessage({
        type: 'request-unhighlight-zone',
        data: {
          zoneId: zoneId
        }
      });
    },
    onScrollToZone: function onScrollToZone(zoneId) {
      sendMessage({
        type: 'request-scroll-to-zone',
        data: {
          zoneId: zoneId
        }
      });
    },
    onPersonificationStart: function onPersonificationStart(userProfile) {
      sendMessage({
        type: 'request-personification-start',
        data: {
          userProfile: userProfile
        }
      });
    },
    onPersonificationEnd: function onPersonificationEnd() {
      sendMessage({
        type: 'request-personification-end',
        data: {}
      });
    }
  };
}
