// react-instantsearch-hooks/dist/cjs/lib/useSearchState.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSearchState = useSearchState;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = require("react");

var _useIndexContext = require("./useIndexContext");

var _useInstantSearchContext = require("./useInstantSearchContext");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function useSearchState() {
  var search = (0, _useInstantSearchContext.useInstantSearchContext)();
  var searchIndex = (0, _useIndexContext.useIndexContext)();
  var indexId = searchIndex.getIndexId();

  var _useState = (0, _react.useState)(function () {
    return search.getUiState();
  }),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      uiState = _useState2[0],
      setLocalUiState = _useState2[1];

  var indexUiState = uiState[indexId];
  var setUiState = (0, _react.useCallback)(function (nextUiState) {
    search.setUiState(nextUiState);
  }, [search]);
  var setIndexUiState = (0, _react.useCallback)(function (nextUiState) {
    setUiState(function (prevUiState) {
      return _objectSpread(_objectSpread({}, prevUiState), {}, (0, _defineProperty2.default)({}, indexId, typeof nextUiState === 'function' ? nextUiState(prevUiState[indexId]) : nextUiState));
    });
  }, [setUiState, indexId]);
  (0, _react.useEffect)(function () {
    function handleRender() {
      setLocalUiState(search.getUiState());
    }

    search.addListener('render', handleRender);
    return function () {
      search.removeListener('render', handleRender);
    };
  }, [search, indexId]);
  return {
    uiState: uiState,
    setUiState: setUiState,
    indexUiState: indexUiState,
    setIndexUiState: setIndexUiState
  };
}
