// react-instantsearch-hooks-web/dist/es/widgets/CurrentRefinements.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["includedAttributes", "excludedAttributes", "transformItems"];
import React from 'react';
import { useCurrentRefinements } from 'react-instantsearch-hooks';
import { CurrentRefinements as CurrentRefinementsUiComponent } from "../ui/CurrentRefinements.js";
export function CurrentRefinements(_ref) {
  var includedAttributes = _ref.includedAttributes,
      excludedAttributes = _ref.excludedAttributes,
      transformItems = _ref.transformItems,
      props = _objectWithoutProperties(_ref, _excluded);

  var _useCurrentRefinement = useCurrentRefinements({
    includedAttributes: includedAttributes,
    excludedAttributes: excludedAttributes,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.currentRefinements'
  }),
      items = _useCurrentRefinement.items,
      refine = _useCurrentRefinement.refine,
      canRefine = _useCurrentRefinement.canRefine;

  var uiProps = {
    items: items,
    onRemove: refine,
    hasRefinements: canRefine
  };
  return /*#__PURE__*/React.createElement(CurrentRefinementsUiComponent, _extends({}, props, uiProps));
}
