// react-instantsearch-hooks-web/dist/cjs/widgets/CurrentRefinements.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CurrentRefinements = CurrentRefinements;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _reactInstantsearchHooks = require("react-instantsearch-hooks");

var _CurrentRefinements = require("../ui/CurrentRefinements");

var _excluded = ["includedAttributes", "excludedAttributes", "transformItems"];

function CurrentRefinements(_ref) {
  var includedAttributes = _ref.includedAttributes,
      excludedAttributes = _ref.excludedAttributes,
      transformItems = _ref.transformItems,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useCurrentRefinement = (0, _reactInstantsearchHooks.useCurrentRefinements)({
    includedAttributes: includedAttributes,
    excludedAttributes: excludedAttributes,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.currentRefinements'
  }),
      items = _useCurrentRefinement.items,
      refine = _useCurrentRefinement.refine,
      canRefine = _useCurrentRefinement.canRefine;

  var uiProps = {
    items: items,
    onRemove: refine,
    hasRefinements: canRefine
  };
  return /*#__PURE__*/_react.default.createElement(_CurrentRefinements.CurrentRefinements, (0, _extends2.default)({}, props, uiProps));
}
