// react-instantsearch-hooks-web/dist/cjs/widgets/HitsPerPage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HitsPerPage = HitsPerPage;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _reactInstantsearchHooks = require("react-instantsearch-hooks");

var _HitsPerPage = require("../ui/HitsPerPage");

var _excluded = ["items", "transformItems"];

function HitsPerPage(_ref) {
  var userItems = _ref.items,
      transformItems = _ref.transformItems,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useHitsPerPage = (0, _reactInstantsearchHooks.useHitsPerPage)({
    items: userItems,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.hitsPerPage'
  }),
      items = _useHitsPerPage.items,
      refine = _useHitsPerPage.refine;

  var _ref2 = items.find(function (_ref3) {
    var isRefined = _ref3.isRefined;
    return isRefined;
  }) || {},
      currentValue = _ref2.value;

  var uiProps = {
    items: items,
    currentValue: currentValue,
    onChange: function onChange(value) {
      return refine(value);
    }
  };
  return /*#__PURE__*/_react.default.createElement(_HitsPerPage.HitsPerPage, (0, _extends2.default)({}, props, uiProps));
}
