// react-instantsearch-hooks-web/dist/cjs/widgets/InfiniteHits.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InfiniteHits = InfiniteHits;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _reactInstantsearchHooks = require("react-instantsearch-hooks");

var _InfiniteHits = require("../ui/InfiniteHits");

var _excluded = ["showPrevious", "cache", "escapeHTML", "showPrevious", "transformItems", "translations"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function InfiniteHits(_ref) {
  var _ref$showPrevious = _ref.showPrevious,
      shouldShowPrevious = _ref$showPrevious === void 0 ? true : _ref$showPrevious,
      cache = _ref.cache,
      escapeHTML = _ref.escapeHTML,
      userShowPrevious = _ref.showPrevious,
      transformItems = _ref.transformItems,
      translations = _ref.translations,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useInfiniteHits = (0, _reactInstantsearchHooks.useInfiniteHits)({
    cache: cache,
    escapeHTML: escapeHTML,
    showPrevious: userShowPrevious,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.infiniteHits'
  }),
      hits = _useInfiniteHits.hits,
      sendEvent = _useInfiniteHits.sendEvent,
      showPrevious = _useInfiniteHits.showPrevious,
      showMore = _useInfiniteHits.showMore,
      isFirstPage = _useInfiniteHits.isFirstPage,
      isLastPage = _useInfiniteHits.isLastPage;

  var uiProps = {
    hits: hits,
    sendEvent: sendEvent,
    onShowPrevious: shouldShowPrevious ? showPrevious : undefined,
    onShowMore: showMore,
    isFirstPage: isFirstPage,
    isLastPage: isLastPage,
    translations: _objectSpread({
      showPreviousButtonText: 'Show previous results',
      showMoreButtonText: 'Show more results'
    }, translations)
  };
  return /*#__PURE__*/_react.default.createElement(_InfiniteHits.InfiniteHits, (0, _extends2.default)({}, props, uiProps));
}
