// react-instantsearch-hooks-web/dist/cjs/widgets/Pagination.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pagination = Pagination;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _reactInstantsearchHooks = require("react-instantsearch-hooks");

var _Pagination = require("../ui/Pagination");

var _excluded = ["showFirst", "showPrevious", "showNext", "showLast", "padding", "totalPages", "translations"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function Pagination(_ref) {
  var showFirst = _ref.showFirst,
      showPrevious = _ref.showPrevious,
      showNext = _ref.showNext,
      showLast = _ref.showLast,
      padding = _ref.padding,
      totalPages = _ref.totalPages,
      translations = _ref.translations,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _usePagination = (0, _reactInstantsearchHooks.usePagination)({
    padding: padding,
    totalPages: totalPages
  }, {
    $$widgetType: 'ais.pagination'
  }),
      pages = _usePagination.pages,
      currentRefinement = _usePagination.currentRefinement,
      isFirstPage = _usePagination.isFirstPage,
      isLastPage = _usePagination.isLastPage,
      nbPages = _usePagination.nbPages,
      createURL = _usePagination.createURL,
      refine = _usePagination.refine;

  var uiProps = {
    pages: pages,
    currentPage: currentRefinement,
    isFirstPage: isFirstPage,
    isLastPage: isLastPage,
    nbPages: nbPages,
    createURL: createURL,
    onNavigate: refine,
    translations: _objectSpread({
      firstPageItemText: '‹‹',
      previousPageItemText: '‹',
      nextPageItemText: '›',
      lastPageItemText: '››',
      pageItemText: function pageItemText(_ref2) {
        var currentPage = _ref2.currentPage;
        return String(currentPage);
      },
      firstPageItemAriaLabel: 'First',
      previousPageItemAriaLabel: 'Previous',
      nextPageItemAriaLabel: 'Next',
      lastPageItemAriaLabel: 'Last',
      pageItemAriaLabel: function pageItemAriaLabel(_ref3) {
        var currentPage = _ref3.currentPage;
        return "Page ".concat(currentPage);
      }
    }, translations)
  };
  return /*#__PURE__*/_react.default.createElement(_Pagination.Pagination, (0, _extends2.default)({}, props, uiProps, {
    showFirst: showFirst,
    showPrevious: showPrevious,
    showNext: showNext,
    showLast: showLast
  }));
}
