// react-instantsearch-hooks-web/dist/cjs/widgets/ToggleRefinement.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToggleRefinement = ToggleRefinement;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _reactInstantsearchHooks = require("react-instantsearch-hooks");

var _ToggleRefinement = require("../ui/ToggleRefinement");

var _excluded = ["attribute", "on", "off"];

function ToggleRefinement(_ref) {
  var attribute = _ref.attribute,
      on = _ref.on,
      off = _ref.off,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useToggleRefinement = (0, _reactInstantsearchHooks.useToggleRefinement)({
    attribute: attribute,
    on: on,
    off: off
  }, {
    $$widgetType: 'ais.toggleRefinement'
  }),
      refine = _useToggleRefinement.refine,
      value = _useToggleRefinement.value;

  var uiProps = {
    checked: value.isRefined,
    onChange: function onChange(isChecked) {
      return refine({
        isRefined: !isChecked
      });
    }
  };
  return /*#__PURE__*/_react.default.createElement(_ToggleRefinement.ToggleRefinement, (0, _extends2.default)({
    label: value.name
  }, props, uiProps));
}
