// @algolia/ui-components-horizontal-slider-react/dist/esm/HorizontalSlider.js
import _extends from "@babel/runtime/helpers/extends";
import { createHorizontalSliderComponent, generateHorizontalSliderId, updateNavigationButtonsProps as _updateNavigationButtonsProps } from '@algolia/ui-components-horizontal-slider-vdom';
import React, { createElement, Fragment, useEffect, useRef } from 'react';
var UncontrolledHorizontalSlider = createHorizontalSliderComponent({
  createElement: createElement,
  Fragment: Fragment
});
export function HorizontalSlider(props) {
  var listRef = useRef(null);
  var nextButtonRef = useRef(null);
  var previousButtonRef = useRef(null);
  var sliderIdRef = useRef(generateHorizontalSliderId());
  useEffect(function () {
    _updateNavigationButtonsProps({
      listRef: listRef,
      nextButtonRef: nextButtonRef,
      previousButtonRef: previousButtonRef
    });
  });
  return /*#__PURE__*/React.createElement(UncontrolledHorizontalSlider, _extends({}, props, {
    listRef: listRef,
    nextButtonRef: nextButtonRef,
    previousButtonRef: previousButtonRef,
    sliderIdRef: sliderIdRef,
    updateNavigationButtonsProps: function updateNavigationButtonsProps() {
      return _updateNavigationButtonsProps({
        listRef: listRef,
        nextButtonRef: nextButtonRef,
        previousButtonRef: previousButtonRef
      });
    }
  }));
}
