// react-instantsearch-hooks-web/dist/es/ui/Menu.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["items", "classNames", "showMore", "canToggleShowMore", "onToggleShowMore", "isShowingMore", "createURL", "onRefine", "translations"];
import React from 'react';
import { cx } from "./lib/cx.js";
import { ShowMoreButton } from "./ShowMoreButton.js";
export function Menu(_ref) {
  var items = _ref.items,
      _ref$classNames = _ref.classNames,
      classNames = _ref$classNames === void 0 ? {} : _ref$classNames,
      showMore = _ref.showMore,
      canToggleShowMore = _ref.canToggleShowMore,
      onToggleShowMore = _ref.onToggleShowMore,
      isShowingMore = _ref.isShowingMore,
      createURL = _ref.createURL,
      onRefine = _ref.onRefine,
      translations = _ref.translations,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement("div", _extends({}, props, {
    className: cx('ais-Menu', classNames.root, items.length === 0 && cx('ais-Menu--noRefinement', classNames.noRefinementRoot), props.className)
  }), /*#__PURE__*/React.createElement("ul", {
    className: cx('ais-Menu-list', classNames.list)
  }, items.map(function (item) {
    return /*#__PURE__*/React.createElement("li", {
      key: item.label,
      className: cx('ais-Menu-item', classNames.item, item.isRefined && cx('ais-Menu-item--selected', classNames.selectedItem))
    }, /*#__PURE__*/React.createElement("a", {
      className: cx('ais-Menu-link', classNames.link),
      href: createURL(item.value),
      onClick: function onClick(event) {
        event.preventDefault();
        onRefine(item);
      }
    }, /*#__PURE__*/React.createElement("span", {
      className: cx('ais-Menu-label', classNames.label)
    }, item.label), /*#__PURE__*/React.createElement("span", {
      className: cx('ais-Menu-count', classNames.count)
    }, item.count)));
  })), showMore && /*#__PURE__*/React.createElement(ShowMoreButton, {
    className: cx('ais-Menu-showMore', classNames.showMore, !canToggleShowMore && cx('ais-Menu-showMore--disabled', classNames.disabledShowMore)),
    disabled: !canToggleShowMore,
    onClick: onToggleShowMore,
    isShowingMore: isShowingMore,
    translations: translations
  }));
}
