// @algolia/predict-core/src/tools/props.ts
import { type SendPredictToolsMessage } from './types';

// @TODO: Installing `@algolia/predict-tools` as a dev dependency to retrieve
// the `PredictToolsProps` type creates Rollup cycle issues:
// "@algolia/predict-core -> @algolia/predict-tools -> @algolia/predict-core"
// We should be able to find a way to install it without cycle issues.
type PredictToolsProps = Record<string, any>;

export function getPredictToolsEventProps(
  sendMessage: SendPredictToolsMessage
): Omit<PredictToolsProps, 'context' | 'version' | 'theme'> {
  return {
    onHighlightZone(zoneId, params) {
      sendMessage({
        type: 'request-highlight-zone',
        data: { zoneId, params },
      });
    },
    onUnhighlightZone(zoneId) {
      sendMessage({
        type: 'request-unhighlight-zone',
        data: { zoneId },
      });
    },
    onScrollToZone(zoneId) {
      sendMessage({
        type: 'request-scroll-to-zone',
        data: { zoneId },
      });
    },
    onPersonificationStart(userProfile) {
      sendMessage({
        type: 'request-personification-start',
        data: { userProfile },
      });
    },
    onPersonificationEnd() {
      sendMessage({
        type: 'request-personification-end',
        data: {},
      });
    },
  };
}
