// react-instantsearch-hooks-web/dist/cjs/ui/Menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Menu = Menu;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _cx = require("./lib/cx");

var _ShowMoreButton = require("./ShowMoreButton");

var _excluded = ["items", "classNames", "showMore", "canToggleShowMore", "onToggleShowMore", "isShowingMore", "createURL", "onRefine", "translations"];

function Menu(_ref) {
  var items = _ref.items,
      _ref$classNames = _ref.classNames,
      classNames = _ref$classNames === void 0 ? {} : _ref$classNames,
      showMore = _ref.showMore,
      canToggleShowMore = _ref.canToggleShowMore,
      onToggleShowMore = _ref.onToggleShowMore,
      isShowingMore = _ref.isShowingMore,
      createURL = _ref.createURL,
      onRefine = _ref.onRefine,
      translations = _ref.translations,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _cx.cx)('ais-Menu', classNames.root, items.length === 0 && (0, _cx.cx)('ais-Menu--noRefinement', classNames.noRefinementRoot), props.className)
  }), /*#__PURE__*/_react.default.createElement("ul", {
    className: (0, _cx.cx)('ais-Menu-list', classNames.list)
  }, items.map(function (item) {
    return /*#__PURE__*/_react.default.createElement("li", {
      key: item.label,
      className: (0, _cx.cx)('ais-Menu-item', classNames.item, item.isRefined && (0, _cx.cx)('ais-Menu-item--selected', classNames.selectedItem))
    }, /*#__PURE__*/_react.default.createElement("a", {
      className: (0, _cx.cx)('ais-Menu-link', classNames.link),
      href: createURL(item.value),
      onClick: function onClick(event) {
        event.preventDefault();
        onRefine(item);
      }
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _cx.cx)('ais-Menu-label', classNames.label)
    }, item.label), /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _cx.cx)('ais-Menu-count', classNames.count)
    }, item.count)));
  })), showMore && /*#__PURE__*/_react.default.createElement(_ShowMoreButton.ShowMoreButton, {
    className: (0, _cx.cx)('ais-Menu-showMore', classNames.showMore, !canToggleShowMore && (0, _cx.cx)('ais-Menu-showMore--disabled', classNames.disabledShowMore)),
    disabled: !canToggleShowMore,
    onClick: onToggleShowMore,
    isShowingMore: isShowingMore,
    translations: translations
  }));
}
