// react-instantsearch-hooks-web/dist/cjs/ui/Hits.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Hits = Hits;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _cx = require("./lib/cx");

var _excluded = ["hits", "sendEvent", "hitComponent", "classNames"];

function DefaultHitComponent(_ref) {
  var hit = _ref.hit;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      wordBreak: 'break-all'
    }
  }, JSON.stringify(hit).slice(0, 100), "\u2026");
}

function Hits(_ref2) {
  var hits = _ref2.hits,
      sendEvent = _ref2.sendEvent,
      _ref2$hitComponent = _ref2.hitComponent,
      HitComponent = _ref2$hitComponent === void 0 ? DefaultHitComponent : _ref2$hitComponent,
      _ref2$classNames = _ref2.classNames,
      classNames = _ref2$classNames === void 0 ? {} : _ref2$classNames,
      props = (0, _objectWithoutProperties2.default)(_ref2, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _cx.cx)('ais-Hits', classNames.root, hits.length === 0 && (0, _cx.cx)('ais-Hits--empty', classNames.emptyRoot), props.className)
  }), /*#__PURE__*/_react.default.createElement("ol", {
    className: (0, _cx.cx)('ais-Hits-list', classNames.list)
  }, hits.map(function (hit) {
    return /*#__PURE__*/_react.default.createElement("li", {
      key: hit.objectID,
      className: (0, _cx.cx)('ais-Hits-item', classNames.item)
    }, /*#__PURE__*/_react.default.createElement(HitComponent, {
      hit: hit,
      sendEvent: sendEvent
    }));
  })));
}
