// react-instantsearch-hooks-web/dist/cjs/ui/InfiniteHits.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InfiniteHits = InfiniteHits;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _cx = require("./lib/cx");

var _excluded = ["hitComponent", "hits", "sendEvent", "isFirstPage", "isLastPage", "onShowPrevious", "onShowMore", "classNames", "translations"];

function DefaultHitComponent(_ref) {
  var hit = _ref.hit;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      wordBreak: 'break-all'
    }
  }, JSON.stringify(hit).slice(0, 100), "\u2026");
}

function InfiniteHits(_ref2) {
  var _ref2$hitComponent = _ref2.hitComponent,
      HitComponent = _ref2$hitComponent === void 0 ? DefaultHitComponent : _ref2$hitComponent,
      hits = _ref2.hits,
      sendEvent = _ref2.sendEvent,
      isFirstPage = _ref2.isFirstPage,
      isLastPage = _ref2.isLastPage,
      onShowPrevious = _ref2.onShowPrevious,
      onShowMore = _ref2.onShowMore,
      _ref2$classNames = _ref2.classNames,
      classNames = _ref2$classNames === void 0 ? {} : _ref2$classNames,
      translations = _ref2.translations,
      props = (0, _objectWithoutProperties2.default)(_ref2, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _cx.cx)('ais-InfiniteHits', classNames.root, hits.length === 0 && (0, _cx.cx)('ais-InfiniteHits--empty', classNames.emptyRoot), props.className)
  }), onShowPrevious && /*#__PURE__*/_react.default.createElement("button", {
    className: (0, _cx.cx)('ais-InfiniteHits-loadPrevious', classNames.loadPrevious, isFirstPage && (0, _cx.cx)('ais-InfiniteHits-loadPrevious--disabled', classNames.disabledLoadPrevious)),
    onClick: onShowPrevious,
    disabled: isFirstPage
  }, translations.showPreviousButtonText), /*#__PURE__*/_react.default.createElement("ol", {
    className: (0, _cx.cx)('ais-InfiniteHits-list', classNames.list)
  }, hits.map(function (hit) {
    return /*#__PURE__*/_react.default.createElement("li", {
      key: hit.objectID,
      className: (0, _cx.cx)('ais-InfiniteHits-item', classNames.item)
    }, /*#__PURE__*/_react.default.createElement(HitComponent, {
      hit: hit,
      sendEvent: sendEvent
    }));
  })), /*#__PURE__*/_react.default.createElement("button", {
    className: (0, _cx.cx)('ais-InfiniteHits-loadMore', classNames.loadMore, isLastPage && (0, _cx.cx)('ais-InfiniteHits-loadMore--disabled', classNames.disabledLoadMore)),
    onClick: onShowMore,
    disabled: isLastPage
  }, translations.showMoreButtonText));
}
