// react-instantsearch-hooks-web/dist/cjs/widgets/Hits.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Hits = Hits;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _reactInstantsearchHooks = require("react-instantsearch-hooks");

var _Hits = require("../ui/Hits");

var _excluded = ["escapeHTML", "transformItems"];

function Hits(_ref) {
  var escapeHTML = _ref.escapeHTML,
      transformItems = _ref.transformItems,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useHits = (0, _reactInstantsearchHooks.useHits)({
    escapeHTML: escapeHTML,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.hits'
  }),
      hits = _useHits.hits,
      sendEvent = _useHits.sendEvent;

  var uiProps = {
    hits: hits,
    sendEvent: sendEvent
  };
  return /*#__PURE__*/_react.default.createElement(_Hits.Hits, (0, _extends2.default)({}, props, uiProps));
}
