// react-instantsearch-hooks-web/dist/es/widgets/Breadcrumb.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["attributes", "rootPath", "separator", "transformItems", "translations"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import React from 'react';
import { useBreadcrumb } from 'react-instantsearch-hooks';
import { Breadcrumb as BreadcrumbUiComponent } from "../ui/Breadcrumb.js";
export function Breadcrumb(_ref) {
  var attributes = _ref.attributes,
      rootPath = _ref.rootPath,
      separator = _ref.separator,
      transformItems = _ref.transformItems,
      translations = _ref.translations,
      props = _objectWithoutProperties(_ref, _excluded);

  var _useBreadcrumb = useBreadcrumb({
    attributes: attributes,
    rootPath: rootPath,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.breadcrumb'
  }),
      canRefine = _useBreadcrumb.canRefine,
      createURL = _useBreadcrumb.createURL,
      items = _useBreadcrumb.items,
      refine = _useBreadcrumb.refine;

  var uiProps = {
    items: items,
    hasItems: canRefine,
    createURL: createURL,
    onNavigate: refine,
    translations: _objectSpread({
      rootElementText: 'Home'
    }, translations)
  };
  return /*#__PURE__*/React.createElement(BreadcrumbUiComponent, _extends({}, props, uiProps));
}
