// react-instantsearch-hooks-web/dist/es/widgets/HitsPerPage.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["items", "transformItems"];
import React from 'react';
import { useHitsPerPage } from 'react-instantsearch-hooks';
import { HitsPerPage as HitsPerPageUiComponent } from "../ui/HitsPerPage.js";
export function HitsPerPage(_ref) {
  var userItems = _ref.items,
      transformItems = _ref.transformItems,
      props = _objectWithoutProperties(_ref, _excluded);

  var _useHitsPerPage = useHitsPerPage({
    items: userItems,
    transformItems: transformItems
  }, {
    $$widgetType: 'ais.hitsPerPage'
  }),
      items = _useHitsPerPage.items,
      refine = _useHitsPerPage.refine;

  var _ref2 = items.find(function (_ref3) {
    var isRefined = _ref3.isRefined;
    return isRefined;
  }) || {},
      currentValue = _ref2.value;

  var uiProps = {
    items: items,
    currentValue: currentValue,
    onChange: function onChange(value) {
      return refine(value);
    }
  };
  return /*#__PURE__*/React.createElement(HitsPerPageUiComponent, _extends({}, props, uiProps));
}
